ChopChop/Models/Recipe/Recipe/Recipe.swift

```diff
@@ -0,0 +1,204 @@
+        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmedName.isEmpty else {
+            throw RecipeError.invalidName
+        }
+        self.name = trimmedName
+
+        guard servings > 0 else {
+            throw RecipeError.invalidServings
+        }
+        self.servings = servings
+        self.difficulty = difficulty
+        self.steps = steps
+        self.ingredients = ingredients
+        assert(checkRepresentation())
+    }
+
+    func updateName(_ name: String) throws {
+        assert(checkRepresentation())
+        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmedName.isEmpty else {
+            throw RecipeError.invalidName
+        }
+        self.name = trimmedName
+        assert(checkRepresentation())
+    }
+
+    func updateRecipeCategory(_ recipeCategory: Int64) {
+        assert(checkRepresentation())
+        self.recipeCategoryId = recipeCategory
+        assert(checkRepresentation())
+    }
+
+    func updateDifficulty(_ difficulty: Difficulty) {
+        assert(checkRepresentation())
+        self.difficulty = difficulty
+        assert(checkRepresentation())
+    }
+
+    // step related functions
+    func addStep(_ addedStep: RecipeStep) {
+        assert(checkRepresentation())
+        steps.append(addedStep)
+        assert(checkRepresentation())
+    }
+
+    func removeStep(_ removedStep: RecipeStep) throws {
+        assert(checkRepresentation())
+        guard (steps.contains { $0 === removedStep }) else {
+            throw RecipeError.nonExistentStep
+        }
+
+        steps.removeAll { $0 === removedStep }
+        assert(checkRepresentation())
+    }
+
+    // reorder step from larger to smaller index
+    func moveStepUp(_ movedStep: RecipeStep) throws {
+        assert(checkRepresentation())
+        guard let idx = (steps.firstIndex { $0 === movedStep }) else {
+            throw RecipeError.nonExistentStep
+        }
+
+        guard idx > 0 else {
+            return
+        }
+
+        steps[idx] = steps[idx - 1]
+        steps[idx - 1] = movedStep
+        assert(checkRepresentation())
+    }
+
+    // reorder step from smaller to larger index
+    func moveStepDown(_ movedStep: RecipeStep) throws {
+        assert(checkRepresentation())
+        guard let idx = (steps.firstIndex { $0 === movedStep }) else {
+            throw RecipeError.nonExistentStep
+        }
+
+        guard idx < steps.count - 1 else {
+            return
+        }
+
+        steps[idx] = steps[idx + 1]
+        steps[idx + 1] = movedStep
+        assert(checkRepresentation())
+    }
+
+    // ingredient related functions
+    func addIngredient(_ addedIngredient: RecipeIngredient) throws {
+        assert(checkRepresentation())
+        guard checkNoDuplicateIngredients(ingredients: ingredients + [addedIngredient]) else {
```

Comment:
Functionality related point: Another way to do this would be to combine the quantity of the `addedIngredient` into the existing one if a duplicate ingredient is found